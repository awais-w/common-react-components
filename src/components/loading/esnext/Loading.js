import React from 'react';
import PropTypes from 'prop-types';
import './Loading.scss';

/**
 * This component displays a simple loading animation
 *
 * @param height height of the component container
 * @param size alter size of the spinner with either sm, md, lg - defaults to lg. For more info: https://www.argos.co.uk/cdn/argos-style/b1.10.06/elements.html#loader
 * @param fullScreen makes the loading component fullscreen
 * @param relativePosition makes component positioning inline
 *
 * @example
 * <Loading
 *    height={200}
 * />
 */
const Loading = (props) => {
  const { height, size, fullScreen, relativePosition, text } = props;
  const sizeClass = size ? `loader-svg-${size}` : '';
  const relativePositionClass = relativePosition ? 'loader-relative' : '';

  return (
    <div role='alert' className={`loading-spinner ${fullScreen ? 'loader-container' : ''}`} style={{ height }}>
      <div className={`loader-svg ${sizeClass} ${relativePositionClass}`}>
        <p className='loader-text'>{text}</p>
        <svg className='circular' viewBox='25 25 50 50'>
          <circle className='path' cx='50' cy='50' r='20' fill='none' />
        </svg>
      </div>
    </div>
  );
};

Loading.propTypes = {
  height: PropTypes.number,
  size: PropTypes.string,
  text: PropTypes.string,
  fullScreen: PropTypes.bool.isRequired,
  relativePosition: PropTypes.bool,
};

Loading.defaultProps = {
  fullScreen: true,
};

export default Loading;
