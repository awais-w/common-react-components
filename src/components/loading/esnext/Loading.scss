@import '@cmc/common-styles/src/scss/_variables.scss';

.loader-container {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: rgba(0, 0, 0, 0.25);
  overflow: hidden;
  z-index: 9999;
  .contained {
    position: absolute;
  }
}

.loader-svg {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  margin: auto;
  width: 12em;
  height: 12em;
  &.loader-relative {
    position: relative;
  }
  &:before {
    position: absolute;
    border: $double solid $color-grey-light;
    border-radius: 50%;
    display: block;
    content: ' ';
    top: $base;
    right: $base;
    bottom: $base;
    left: $base;
  }
  &.loader-svg-lg {
    width: $double * 12;
    height: $double * 12;
  }
  &.loader-svg-md {
    width: $double * 8;
    height: $double * 8;
    &:before {
      border-width: $medium;
    }
  }
  &.loader-svg-sm {
    width: $double * 6;
    height: $double * 6;
    &:before {
      border-width: $base;
    }
  }
  .loader-text {
    position: absolute;
    width: 100%;
    bottom: -$large;
    text-align-last: center;
    @include font-standard;
  }
  .circular {
    animation: rotater 3s linear infinite;
    transform-origin: center center;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    margin: auto;
  }
  .path {
    stroke-dasharray: 10, 200;
    stroke-dashoffset: 0;
    stroke-width: 0.2em;
    stroke-linecap: round;
    stroke: $primary-color;
    animation: dash 3s ease-in-out infinite;
  }
}

@keyframes rotater {
  100% {
    transform: rotate(360deg);
  }
}

@-webkit-keyframes rotater {
  100% {
    -webkit-transform: rotate(360deg);
  }
}

@keyframes dash {
  0% {
    stroke-dasharray: 1, 200;
    stroke-dashoffset: 0;
  }
  50% {
    stroke-dasharray: 89, 200;
    stroke-dashoffset: -35px;
  }
  100% {
    stroke-dasharray: 125, 200;
    stroke-dashoffset: -124px;
  }
}
