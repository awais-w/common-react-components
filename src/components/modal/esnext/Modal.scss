@import '@cmc/common-styles/src/scss/_variables.scss';
.cmc-modal {
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
  background: rgba(0, 0, 0, 0.6);
  z-index: 1000;
  @include font-standard;
  display: flex;
  align-items: center;
  justify-content: center;

  &__container {
    width: 50%;
    padding: 0;
    border-radius: $half;
    background: $color-white;
    position: relative;
    z-index: 6;
    overflow: hidden;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    @include prefix(box-shadow, 0px 0px $large $half rgba(0, 0, 0, 0.7), webkit moz);

    &-header {
      display: flex;
      align-items: center;
      padding: $base $double;
      margin: 0;
      overflow: hidden;
      border-bottom: 1px solid $color-grey-mid;
      @include font-condensed;

      h1 {
        margin: 0;
        padding: 0;
        text-transform: uppercase;
        font-size: $font-size-h2;
      }

      #close {
        align-self: center;
        justify-self: flex-end;
        order: 2;
        margin-left: auto;
        margin-top: 0;
        font-size: 35px;
        cursor: pointer;
      }
    }

    &-content {
      width: 100%;
      margin: $double 0;
      padding: 0 $double;
      text-align: center;
      max-height: calc(100vh - 180px);
      overflow: auto;
    }

    &-footer {
      display: flex;
      justify-content: center;
      text-transform: uppercase;
      padding: 0 $double;
      margin: $double 0;
      &-button:not(:last-child) {
        margin-right: $double;
      }
    }
  }

  &__outside {
    width: 100%;
    height: 100%;
    z-index: 5;
    position: absolute;
    top: 0;
  }
}
