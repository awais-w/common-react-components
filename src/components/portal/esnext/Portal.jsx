import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

const Portal = ({ children, nodeSelector }) => {
  try {
    return ReactDOM.createPortal(children, document.querySelector(nodeSelector));
  } catch (error) {
    console.error(`Could not find a matching node for specified node selector. The element is rendered at the body.`);
    return ReactDOM.createPortal(children, document.querySelector('body'));
  }
};

Portal.propTypes = {
  children: PropTypes.any,
  nodeSelector: PropTypes.string,
};

Portal.defaultProps = {
  nodeSelector: 'body',
};

export default Portal;
