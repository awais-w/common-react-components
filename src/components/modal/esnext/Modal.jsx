import React from 'react';
import PropTypes from 'prop-types';
import bem from '@argos/utils/esnext/bem';
import Button from '@cmc/cha-button/esnext/Button';
import Icon from '@cmc/icon/esnext/Icon';
import Loading from '@cmc/loading/esnext/Loading';
import Portal from '@cmc/portal/esnext/Portal';
import './Modal.scss';

const propTypes = {
  onClose: PropTypes.func,
  parentSelector: PropTypes.string,
  primaryButton: PropTypes.shape({
    handleClick: PropTypes.func,
    isDisabled: PropTypes.bool,
    isLoading: PropTypes.bool,
    text: PropTypes.string,
  }),
  secondaryButton: PropTypes.shape({
    handleClick: PropTypes.func,
    isDisabled: PropTypes.bool,
    text: PropTypes.string,
  }),
  children: PropTypes.node,
  title: PropTypes.string,
  isLoading: PropTypes.bool,
  width: PropTypes.string,
  minWidth: PropTypes.string,
  height: PropTypes.string,
  minHeight: PropTypes.string,
};

const cn = bem({ prefix: 'cmc-', block: 'modal' });

const Modal = ({
  onClose,
  title,
  children,
  primaryButton,
  secondaryButton,
  isLoading,
  parentSelector,
  width,
  minWidth,
  height,
  minHeight,
  classes,
}) => {
  const shouldShowHeader = !isLoading && (onClose || title);

  const shouldShowFooter = !isLoading && (secondaryButton || primaryButton);

  return (
    <Portal nodeSelector={parentSelector}>
      <div className={cn(null, null, classes)}>
        <div className={cn('container')} style={{ minWidth, width, minHeight, height }}>
          {shouldShowHeader && (
            <div className={cn('container-header')}>
              {onClose && <Icon id={'close'} icon='close' onClick={onClose} />}
              {title && <h1>{title}</h1>}
            </div>
          )}
          <div className={cn('container-content')}>
            {isLoading ? <Loading fullScreen={false} height={200} /> : children}
          </div>
          {shouldShowFooter && (
            <div className={cn('container-footer')}>
              {secondaryButton && (
                <div className={cn('container-footer-button')}>
                  <Button
                    text={secondaryButton.text}
                    onClick={secondaryButton.handleClick}
                    isDisabled={secondaryButton.isDisabled}
                    buttonWidth='half'
                    isSecondary
                    isShort
                  />
                </div>
              )}
              {primaryButton && (
                <div className={cn('container-footer-button')}>
                  <Button
                    text={primaryButton.text}
                    loadingText={primaryButton.loadingText}
                    onClick={primaryButton.handleClick}
                    buttonWidth='half'
                    isDisabled={primaryButton.isDisabled}
                    isLoading={primaryButton.isLoading}
                    isShort
                  />
                </div>
              )}
            </div>
          )}
        </div>
        <div className={cn('outside')} />
      </div>
    </Portal>
  );
};

Modal.propTypes = propTypes;
export default Modal;
