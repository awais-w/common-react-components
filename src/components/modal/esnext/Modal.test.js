import React from 'react';
import { shallow } from 'enzyme';
import Loading from '@cmc/loading/esnext/Loading';
import Portal from '@cmc/portal/esnext/Portal';
import Modal from './Modal';

const defaultProps = {
  onClose: jest.fn(),
  primaryButton: { handleClick: jest.fn(), text: 'primary', isDisabled: false },
  secondaryButton: { handleClick: jest.fn(), text: 'secondary', isDisabled: false },
  children: 'Modal content',
  title: 'Modal title',
};

describe('Modal', () => {
  const setup = ({ props }) => {
    const component = shallow(<Modal {...props} />);
    return {
      wrapper: component,
      getPortal: () => component.find(Portal),
      getHeader: () => component.find('.cmc-modal__container-header'),
      getCloseButton: () => component.find('#close'),
      getContent: () => component.find('.cmc-modal__container-content'),
      getPrimaryButton: () =>
        component.findWhere((element) => element.name() === 'Button' && !element.prop('isSecondary')),
      getSecondaryButton: () =>
        component.findWhere((element) => element.name() === 'Button' && element.prop('isSecondary')),
      getLoading: () => component.find(Loading),
      getFooter: () => component.find('.cmc-modal__container-footer'),
      getOutside: () => component.find('.cmc-modal__outside'),
    };
  };

  describe('base tests', () => {
    const component = setup({ props: defaultProps });
    it('exists', () => {
      expect(component.wrapper).toExist();
    });
    it('should match default snapshot', () => {
      expect(component.wrapper).toMatchSnapshot();
    });
    it('should render correct content', () => {
      expect(component.getCloseButton()).toHaveLength(1);
      expect(component.getHeader().text()).toContain(defaultProps.title);
      expect(component.getContent().text()).toEqual(defaultProps.children);
      expect(component.getPrimaryButton()).toHaveLength(1);
      expect(component.getSecondaryButton()).toHaveLength(1);
    });
    it('renders a portal component', () => {
      expect(component.getPortal()).toExist();
    });
  });

  describe('conditional renders', () => {
    it('does not render header if props not provided', () => {
      const component = setup({ props: { ...defaultProps, title: '', onClose: null } });
      expect(component.getHeader()).toHaveLength(0);
    });

    it('does not render footer if props not provided', () => {
      const component = setup({ props: { ...defaultProps, primaryButton: null, secondaryButton: null } });
      expect(component.getFooter()).toHaveLength(0);
    });

    it('renders loading component is isLoading is set to true', () => {
      const component = setup({ props: { ...defaultProps, isLoading: true } });
      expect(component.getLoading()).toHaveLength(1);
    });

    describe('when secondary button not provided', () => {
      const component = setup({ props: { ...defaultProps, secondaryButton: null } });
      it('renders footer', () => {
        expect(component.getFooter()).toHaveLength(1);
      });
      it('does not render a secondary button', () => {
        expect(component.getSecondaryButton()).toHaveLength(0);
      });
    });

    describe('when primary button not provided', () => {
      const component = setup({ props: { ...defaultProps, primaryButton: null } });
      it('renders footer', () => {
        expect(component.getFooter()).toHaveLength(1);
      });
      it('does not render a primary button', () => {
        expect(component.getPrimaryButton()).toHaveLength(0);
      });
    });
  });

  describe('button clicks', () => {
    const component = setup({ props: defaultProps });
    it('calls onClose when close button is clicked', () => {
      component.getCloseButton().simulate('click');
      expect(defaultProps.onClose).toBeCalledTimes(1);
    });

    it('calls primaryButton onclick when primary button is clicked', () => {
      component.getPrimaryButton().simulate('click');
      expect(defaultProps.primaryButton.handleClick).toBeCalledTimes(1);
    });

    it('calls secondaryButton onclick when secondary button is clicked', () => {
      component.getSecondaryButton().simulate('click');
      expect(defaultProps.secondaryButton.handleClick).toBeCalledTimes(1);
    });
  });

  describe('Custom classes', () => {
    const component = setup({ props: { ...defaultProps, classes: 'class1 class2' } });
    it('adds custom classes', () => {
      expect(component.wrapper.find('.class1')).toHaveLength(1);
      expect(component.wrapper.find('.class2')).toHaveLength(1);
    });
  });
});
