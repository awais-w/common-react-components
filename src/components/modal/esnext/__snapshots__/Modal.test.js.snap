// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal base tests should match default snapshot 1`] = `
<Portal
  nodeSelector="body"
>
  <div
    className="cmc-modal"
  >
    <div
      className="cmc-modal__container"
      style={
        Object {
          "height": undefined,
          "minHeight": undefined,
          "minWidth": undefined,
          "width": undefined,
        }
      }
    >
      <div
        className="cmc-modal__container-header"
      >
        <Icon
          icon="close"
          id="close"
          isCustom={false}
          onClick={[MockFunction]}
        />
        <h1>
          Modal title
        </h1>
      </div>
      <div
        className="cmc-modal__container-content"
      >
        Modal content
      </div>
      <div
        className="cmc-modal__container-footer"
      >
        <div
          className="cmc-modal__container-footer-button"
        >
          <Button
            buttonWidth="half"
            icon=""
            iconMode="left"
            isDisabled={false}
            isLoading={false}
            isSecondary={true}
            isShort={true}
            loadingText=""
            onClick={[MockFunction]}
            text="secondary"
            type="button"
          />
        </div>
        <div
          className="cmc-modal__container-footer-button"
        >
          <Button
            buttonWidth="half"
            icon=""
            iconMode="left"
            isDisabled={false}
            isLoading={false}
            isSecondary={false}
            isShort={true}
            loadingText=""
            onClick={[MockFunction]}
            text="primary"
            type="button"
          />
        </div>
      </div>
    </div>
    <div
      className="cmc-modal__outside"
    />
  </div>
</Portal>
`;
